File: python/twelve-days/twelve_days.py
"""
Module providing a function to recite verses of "The Twelve Days of Christmas".
"""

LYRICS = {
    1: ("first", "a Partridge in a Pear Tree."),
    2: ("second", "two Turtle Doves, and "),
    3: ("third", "three French Hens, "),
    4: ("fourth", "four Calling Birds, "),
    5: ("fifth", "five Gold Rings, "),
    6: ("sixth", "six Geese-a-Laying, "),
    7: ("seventh", "seven Swans-a-Swimming, "),
    8: ("eighth", "eight Maids-a-Milking, "),
    9: ("ninth", "nine Ladies Dancing, "),
    10: ("tenth", "ten Lords-a-Leaping, "),
    11: ("eleventh", "eleven Pipers Piping, "),
    12: ("twelfth", "twelve Drummers Drumming, ")
}

def recite(start_verse, end_verse):
    """
    Returns a list of verses from start_verse to end_verse.
    """
    if start_verse <= 0 or end_verse <= 0 or start_verse > end_verse:
        return []
        
    result = []
    for day_index in range(start_verse, end_verse + 1):
        current_verse = "On the " + LYRICS[day_index][0] + " day of Christmas my true love gave to me: "
        for gift_index in range(day_index, 0, -1):
            current_verse += LYRICS[gift_index][1]
        result.append(current_verse)

    return result

File: cpp/twelve-days/twelve_days.cpp
#include "twelve_days.h"

#include <array>
#include <string>

namespace twelve_days {

const std::array<std::string, 12> DAYS = {
    "first", "second", "third", "fourth", "fifth", "sixth",
    "seventh", "eighth", "ninth", "tenth", "eleventh", "twelfth"
};

const std::array<std::string, 12> GIFTS = {
    "a Partridge in a Pear Tree.\n",
    "two Turtle Doves, and ",
    "three French Hens, ",
    "four Calling Birds, ",
    "five Gold Rings, ",
    "six Geese-a-Laying, ",
    "seven Swans-a-Swimming, ",
    "eight Maids-a-Milking, ",
    "nine Ladies Dancing, ",
    "ten Lords-a-Leaping, ",
    "eleven Pipers Piping, ",
    "twelve Drummers Drumming, "
};

std::string recite(int start_verse, int end_verse)
{
    if(start_verse <= 0 || end_verse <= 0 || start_verse > end_verse)
        return "";

    const int start_day = start_verse - 1;
    const int end_day = end_verse - 1;
    
    std::string result = "";
    for(int day_index = start_day; day_index <= end_day; day_index++)
    {
        std::string current_verse = "On the " + DAYS[day_index] + " day of Christmas my true love gave to me: ";
        for(int gift_index = day_index; gift_index >= 0; gift_index--)
        {
            current_verse += GIFTS[gift_index];
        }

        if(day_index < end_day) {
            current_verse += '\n';
        }

        result += current_verse;
    }

    return result;
}

}  // namespace twelve_days
